src/lib/arch/aarch64/Instruction_decode.cc

```diff
@@ -660,36 +831,42 @@ void Instruction::decode() {
     setInstructionType(InsnType::isPredicate);
   }
   // Uncaught float data assignment for FMOV move to general instructions
-  if (((430 <= metadata_.opcode && metadata_.opcode <= 432) ||
-       (2409 <= metadata_.opcode && metadata_.opcode <= 2429)) &&
+  if (((Opcode::AArch64_FMOVD0 <= metadata_.opcode &&
+        metadata_.opcode <= Opcode::AArch64_FMOVS0) ||
+       (Opcode::AArch64_FMOVDXHighr <= metadata_.opcode &&
+        metadata_.opcode <= Opcode::AArch64_FMOVv8f16_ns)) &&
       !(isInstruction(InsnType::isScalarData) ||
         isInstruction(InsnType::isVectorData))) {
     setInstructionType(InsnType::isScalarData);
   }
   // Uncaught vector data assignment for SMOV and UMOV instructions
-  if ((4341 <= metadata_.opcode && metadata_.opcode <= 4350) ||
-      (5795 <= metadata_.opcode && metadata_.opcode <= 5802)) {
+  if ((Opcode::AArch64_SMOVvi16to32 <= metadata_.opcode &&
+       metadata_.opcode <= Opcode::AArch64_SMOVvi8to64_idx0) ||
+      (Opcode::AArch64_UMOVvi16 <= metadata_.opcode &&
+       metadata_.opcode <= Opcode::AArch64_UMOVvi8_idx0)) {
     setInstructionType(InsnType::isVectorData);
   }
   // Uncaught float data assignment for FCVT convert to general instructions
-  if ((1976 <= metadata_.opcode && metadata_.opcode <= 2186) &&
+  if ((Opcode::AArch64_FCVTASUWDr <= metadata_.opcode &&
+       metadata_.opcode <= Opcode::AArch64_FCVT_ZPmZ_StoH) &&
       !(isInstruction(InsnType::isScalarData) ||
         isInstruction(InsnType::isVectorData))) {
     setInstructionType(InsnType::isScalarData);
   }
 
-  if (!(isInstruction(InsnType::isSMEData))) {
-    // Catch zero register references and pre-complete those operands - not
-    // applicable to SME instructions
-    for (uint16_t i = 0; i < sourceRegisterCount_; i++) {
-      if (sourceRegisters_[i] == RegisterType::ZERO_REGISTER) {
-        sourceValues_[i] = RegisterValue(0, 8);
-        sourceOperandsPending_--;
-      }
+  // if (!(isInstruction(InsnType::isSMEData))) {
+  // Catch zero register references and pre-complete those operands - not
+  // applicable to SME instructions
+  for (uint16_t i = 0; i < sourceRegisterCount_; i++) {
+    if (sourceRegisters_[i] == RegisterType::ZERO_REGISTER) {
+      sourceValues_[i] = RegisterValue(0, 8);
+      sourceOperandsPending_--;
     }
-  } else {
-    // For SME instructions, resize the following structures to have the exact
-    // amount of space required
+  }
+  // } else {
+  if (isInstruction(InsnType::isSMEData)) {
```

Comment:
To remind myself, for SME, do we expand these containers to their maximum and then shrink them back down to the actual usage?

Reply:
Correct yes